### PR TITLE
Nrf_Cache/Nrf_Subscribe_Notify Feature

### PR DESCRIPTION
**Pull request by aditya-mehta14**
This PR involves use of NRF Cache Feature and NRF Subscribe/Notify service from nrfcache package in NRF.

### CODE REVIEW

File: producer/callback.go [hunk 1]
```diff
@@ -126,3 +130,52 @@ func BuildAndSendQosN1N2TransferMsg(smContext *smf_context.SMContext) error {
...
+			Detail: "Missing IE [Event]/[NfInstanceUri] in NotificationData",
+		}
+		return problemDetails
+	}
+	nfInstanceId := notificationData.NfInstanceUri[strings.LastIndex(notificationData.NfInstanceUri, "/")+1:]
+
+	// If nrf caching is enabled, go ahead and delete the entry from the cache.
+	// This will force the smf to do nf discovery and get the updated nf profile from the nrf.
+	if smf_context.SMF_Self().EnableNrfCaching {
+		ok := nrf_cache.RemoveNfProfileFromNrfCache(nfInstanceId)
```
Comment (UpadhyayaAnkur) [inline producer/callback.go:163]: why notification is removal of profile ? Can you please explain with detailed comment of logic

File: producer/callback.go [hunk 2]
```diff
@@ -126,3 +130,52 @@ func BuildAndSendQosN1N2TransferMsg(smContext *smf_context.SMContext) error {
...
+		logger.PduSessLog.Tracef("nfinstance %v deleted from cache: %v", nfInstanceId, ok)
+	}
+	if subscriptionId, ok := smf_context.SMF_Self().NfStatusSubscriptions.Load(nfInstanceId); ok {
+		logger.PduSessLog.Debugf("SubscriptionId of nfInstance %v is %v",nfInstanceId,subscriptionId.(string))
+		problemDetails, err := consumer.SendRemoveSubscription(subscriptionId.(string))
```
Comment (UpadhyayaAnkur) [inline producer/callback.go:168]: Sending http request, within handling of ongoing http request, back to same peer may not be good idea. Pl check.
Here NRF has sent notification to SMF on de-registration of some NF and SMF is initiating Remove_Subscription back to NRF within handling of notification from NRF.

File: service/init.go [hunk 3]
```diff
@@ -318,6 +319,11 @@ func (smf *SMF) Start() {
 	//Send NRF Registration
 	smf.SendNrfRegistration()
 
+	if smfCtxt.EnableNrfCaching {
+		initLog.Infoln("Enable NRF caching feature")
```
Comment (UpadhyayaAnkur) [inline service/init.go:323]: Probably, print cache eviction time from config also in this log

File: factory/config.go [hunk 4]
```diff
@@ -71,21 +71,23 @@ type KafkaInfo struct {
...
+	UserPlaneInformation     UserPlaneInformation `yaml:"userplane_information"`
+	ServiceNameList          []string             `yaml:"serviceNameList,omitempty"`
+	SNssaiInfo               []SnssaiInfoItem     `yaml:"snssaiInfos,omitempty"`
+	ULCL                     bool                 `yaml:"ulcl,omitempty"`
+	EnterpriseList           map[string]string    `yaml:"enterpriseList,omitempty"`
+	EnableDbStore            bool                 `yaml:"enableDBStore,omitempty"`
+	EnableUpfAdapter         bool                 `yaml:"enableUPFAdapter,omitempty"`
+	SmfDbName                string               `yaml:"smfDBName,omitempty"`
+	Mongodb                  *Mongodb             `yaml:"mongodb,omitempty"`
+	EnableNrfCaching         bool                 `yaml:"enableNrfCaching"`
```
Comment (UpadhyayaAnkur) [inline factory/config.go:87]: Grouping NRF cache related config will look clean

File: context/context.go [hunk 5]
```diff
@@ -70,13 +72,16 @@ type SMFContext struct {
...
-	DrsmCtxts           DrsmCtxts
-
-	EnterpriseList *map[string]string // map to contain slice-name:enterprise-name
-	PodIp          string
+	ULCLSupport              bool
+	UEPreConfigPathPool      map[string]*UEPreConfigPaths
+	LocalSEIDCount           uint64
+	DrsmCtxts                DrsmCtxts
+	EnterpriseList           *map[string]string // map to contain slice-name:enterprise-name
+	EnableNrfCaching         bool
```
Comment (UpadhyayaAnkur) [inline context/context.go:80]: grouping NRF cache config will look clean

File: consumer/nnrf.go [hunk 6]
```diff
@@ -348,3 +380,52 @@ func SendDeregisterNFInstance() (*models.ProblemDetails, error) {
...
+		problem := err.(openapi.GenericOpenAPIError).Model().(models.ProblemDetails)
+		problemDetails = &problem
+	} else {
+		err = openapi.ReportError("server no response")
+	}
+	return
+}
+
+func SendRemoveSubscription(subscriptionId string) (problemDetails *models.ProblemDetails, err error) {
+	logger.ConsumerLog.Infoln("[SMF] Send Remove Subscription")
```
Comment (UpadhyayaAnkur) [inline consumer/nnrf.go:410]: Add what subscription ID is being removed in the log and if possible then NF type too.

File: consumer/nnrf.go [hunk 7]
```diff
@@ -348,3 +380,52 @@ func SendDeregisterNFInstance() (*models.ProblemDetails, error) {
 		return nil, openapi.ReportError("server no response")
 	}
 }
+
+func SendCreateSubscription(nrfUri string, nrfSubscriptionData models.NrfSubscriptionData) (nrfSubData models.NrfSubscriptionData, problemDetails *models.ProblemDetails, err error) {
+	logger.ConsumerLog.Debugf("Send Create Subscription")
```
Comment (UpadhyayaAnkur) [inline consumer/nnrf.go:385]: Adding what NF is being subscribed with some details shall be helpful in the log

File: callback/api_nf_subscribe_notify.go [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+		return
+	}
+
+	req := http_wrapper.NewRequest(c.Request, nfSubscriptionStatusNotification)
+
+	rsp := producer.HandleNfSubscriptionStatusNotify(req)
+
+	responseBody, err := openapi.Serialize(rsp.Body, "application/json")
+	if err != nil {
+		logger.PduSessLog.Errorln(err)
```
Comment (UpadhyayaAnkur) [inline callback/api_nf_subscribe_notify.go:50]: Error log with some content shall be helpful

File: producer/callback.go (hunk 1 above)
Comment (aditya-mehta14) [reply producer/callback.go:162]: As soon as a notification regarding a NF profile change is sent, the old NF profile is removed from the cache and an updated NF Profile is fetched in the next discovery request.

File: producer/callback.go (hunk 2 above)
Comment (aditya-mehta14) [reply producer/callback.go]: Modified the logic.

File: service/init.go (hunk 3 above)
Comment (aditya-mehta14) [reply service/init.go]: Done.

File: consumer/nnrf.go (hunk 6 above)
Comment (aditya-mehta14) [reply consumer/nnrf.go]: Added subscription ID in the log.

File: consumer/nnrf.go (hunk 7 above)
Comment (aditya-mehta14) [reply consumer/nnrf.go]: Added NF type.

**Review by badhrinathpa — APPROVED** [review]
+1

File: context/context.go (hunk 5 above)
Comment (aditya-mehta14) [reply context/context.go]: Done.

File: callback/api_nf_subscribe_notify.go (hunk 8 above)
Comment (aditya-mehta14) [reply callback/api_nf_subscribe_notify.go]: Added

**Review by UpadhyayaAnkur — APPROVED** [review]

**Review by UpadhyayaAnkur — APPROVED** [review]

**Comment by amolonf** [comment]
retest this please

**Review by amolonf — COMMENTED** [review]
Added license information to api_nf_subscribe_notify.go

**Comment by amolonf** [comment]
retest this please

**Review by UpadhyayaAnkur — APPROVED** [review]

**Pull request merged**